CalibTracker/SiStripESProducers/plugins/real/SiStripQualityESProducer.cc

```diff
@@ -16,92 +16,153 @@
-      quality->add( obj.product() );
+      badStripGetters_.emplace_back(make_badStripGetter<SiStripBadModuleRcd>(cc, tagName));
     } else if (recordName=="SiStripBadModuleFedErrRcd"){
-      iRecord.getRecord<SiStripBadModuleFedErrRcd>().get(tagName,obj); 
-      quality->add( obj.product() );
+      badStripGetters_.emplace_back(make_badStripGetter<SiStripBadModuleFedErrRcd>(cc, tagName));
     } else if (recordName=="SiStripBadFiberRcd"){
-      iRecord.getRecord<SiStripBadFiberRcd>().get(tagName,obj); 
-      quality->add( obj.product() );    
+      badStripGetters_.emplace_back(make_badStripGetter<SiStripBadFiberRcd>(cc, tagName));
     } else if (recordName=="SiStripBadChannelRcd"){
-      iRecord.getRecord<SiStripBadChannelRcd>().get(tagName,obj);
-      quality->add( obj.product() );    
+      badStripGetters_.emplace_back(make_badStripGetter<SiStripBadChannelRcd>(cc, tagName));
     } else if (recordName=="SiStripBadStripRcd"){
-      iRecord.getRecord<SiStripBadStripRcd>().get(tagName,obj); 
-      quality->add( obj.product() );    
+      badStripGetters_.emplace_back(make_badStripGetter<SiStripBadStripRcd>(cc, tagName));
     } else if (recordName=="SiStripDetCablingRcd"){
-      iRecord.getRecord<SiStripDetCablingRcd>().get(tagName,cabling);
-      quality->add( cabling.product() );
+      cablingGetters_.emplace_back(cc, tagName);
     } else if (recordName=="SiStripDetVOffRcd"){
-      iRecord.getRecord<SiStripDetVOffRcd>().get(tagName,Voff);
-      quality->add( Voff.product() );
+      voffGetters_.emplace_back(cc, tagName);
     } else if (recordName=="RunInfoRcd") {
       runInfoTagName = tagName;
       doRunInfo = true;
     } else {
-      edm::LogError("SiStripQualityESProducer") << "[SiStripQualityESProducer::produce] Skipping the requested data for unexisting record " << recordName << " with tag " << tagName << std::endl;
+      // Would it make sense to elevate this to an exception?
+      edm::LogError("SiStripQualityESProducer") << "[SiStripQualityESProducer::ctor] Skipping the requested data for unexisting record " << recordName << " with tag " << tagName << std::endl;
       continue;
     }
   }
+
+  if(doRunInfo) {
+    cc.setConsumes(runInfoToken_, edm::ESInputTag{"", runInfoTagName});
+  }
+}
+
+
+std::unique_ptr<SiStripQuality> SiStripQualityESProducer::produce(const SiStripQualityRcd& iRecord)
+{
+  auto quality = std::make_unique<SiStripQuality>();  
+  edm::LogInfo("SiStripQualityESProducer") << "produce called";
+
+  // Set the debug output level
+  quality->setPrintDebugOutput(printDebugOutput_);
+  // Set the protection against empty RunInfo objects
+  quality->setUseEmptyRunInfo(useEmptyRunInfo_);
+
+  for(const auto& getter: badStripGetters_) {
+    quality->add(&getter->get(iRecord));
+  }
+  for(const auto& getter: cablingGetters_) {
+    quality->add(&getter.get(iRecord));
+  }
+  for(const auto& getter: voffGetters_) {
+    quality->add(&getter.get(iRecord));
+  }
```

Comment:
Previously the objects were added to `SiStripQuality` in the order of configuration. Now the order is first all `SiStripBadStrip` objects, then all cabling objects, and then all VOff objects. Does the order matter? (if it does, I'll change the code accordingly)

Reply:
Hi @makortel I would prefer if you could maintain the order as it was implemented. 
I've seen that when merging the bad components from different sources to create a Strip bad components DB scenario it can actually lead to different results.

Reply:
Thanks @mmusich, I'll change the order back. Could someone add the hold back?

Reply:
The last commit should now preserve the order.